src/rust/src/builder_ffi.rs

```diff
@@ -0,0 +1,169 @@
+use std::convert::TryInto;
+use std::ptr;
+
+use orchard::{
+    builder::{Builder, InProgress, Unauthorized, Unproven},
+    bundle::{Authorized, Flags},
+    keys::OutgoingViewingKey,
+    value::NoteValue,
+    Bundle,
+};
+use rand_core::OsRng;
+use tracing::error;
+use zcash_primitives::transaction::{
+    components::{sapling, transparent, Amount},
+    sighash::{SignableInput, SIGHASH_ALL},
+    sighash_v5::v5_signature_hash,
+    Authorization, TransactionData, TxVersion,
+};
+
+use crate::{transaction_ffi::PrecomputedTxParts, ORCHARD_PK};
+
+#[no_mangle]
+pub extern "C" fn orchard_builder_new(
+    spends_enabled: bool,
+    outputs_enabled: bool,
+    anchor: *const [u8; 32],
+) -> *mut Builder {
+    let anchor = unsafe { anchor.as_ref() }.expect("Anchor pointer may not be null.");
+    Box::into_raw(Box::new(Builder::new(
+        Flags::from_parts(spends_enabled, outputs_enabled),
+        orchard::Anchor::from_bytes(*anchor).unwrap(),
+    )))
+}
+
+#[no_mangle]
+pub extern "C" fn orchard_builder_add_recipient(
+    builder: *mut Builder,
+    ovk: *const [u8; 32],
+    recipient: *const orchard::Address,
+    value: u64,
+    memo: *const [u8; 512],
+) -> bool {
+    let builder = unsafe { builder.as_mut() }.expect("Builder may not be null.");
+    let ovk = unsafe { ovk.as_ref() }
+        .copied()
+        .map(OutgoingViewingKey::from);
+    let recipient = unsafe { recipient.as_ref() }.expect("Recipient may not be null.");
+    let value = NoteValue::from_raw(value);
+    let memo = unsafe { memo.as_ref() }.expect("Memo may not be null.");
```

Comment:
`TransactionBuilder::AddOrchardOutput` was emulating the `AddSaplingOutput` API, which uses a default `NO_MEMO` instead of a `std::optional`. But yeah, I could refactor this to instead have the no-memo variant be handled in Rust.